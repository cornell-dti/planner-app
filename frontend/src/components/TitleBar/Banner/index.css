.Banner {
  position: fixed;
  align-items: center;
  align-content: center;
  top: 0;
  left: 0;
  right: 0;
  z-index: 20;
  height: auto;
  display: flex;
  color: white;
  padding: 1em;
  background-color: #7fbdff;
}

.Text {
  flex: auto;
}

.DismissButton {
  color: white;
  background: none;
  padding: 8px;
  border: 1px solid white;
  border-radius: 5px;
  cursor: pointer;
}
