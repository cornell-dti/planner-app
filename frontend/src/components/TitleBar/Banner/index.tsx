import React, { ReactElement } from 'react';
import { connect } from 'react-redux';
import { State } from 'common/lib/types/store-types';
import { getBannerMessage } from '../../../store/selectors';
import styles from './index.module.css';
import { readBannerMessage } from '../../../firebase/actions';
import { MessageWithId } from './messages';

type Props = {
  readonly message: MessageWithId | null;
};

const Banner = ({ message }: Props): ReactElement | null => {
  if (message === null) {
    return null;
  }
  const dismissHandler = (): void => {
    readBannerMessage(message.id, true);
  };
  return (
    <div className={styles.Banner}>
      <span className={styles.Text}>{message.message}</span>
      <button type="button" className={styles.DismissButton} onClick={dismissHandler}>
        Dismiss
      </button>
    </div>
  );
};

const Connected = connect((state: State) => getBannerMessage(state))(Banner);
export default Connected;
