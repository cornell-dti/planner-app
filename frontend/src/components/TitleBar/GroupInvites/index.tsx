import React, { ReactElement } from 'react';
import { connect } from 'react-redux';
import { Map } from 'immutable';
import { State, PendingGroupInvite } from 'common/lib/types/store-types';
import styles from './index.module.css';
import { rejectInvite } from '../../../firebase/actions';

type Props = {
  readonly pendingInvites: Map<string, PendingGroupInvite>;
};

function SingleInvitation(
  invite: PendingGroupInvite,
): ReactElement {
  return (
    <li key={invite.group}>
      <span className={styles.Text}>
        {`${invite.inviterName} has invited you to join their group project.`}
      </span>
      <div className={styles.ButtonWrap}>
        <button
          type="button"
          onClick={() => {
            console.log(`Tried to join group ${invite.group}, a currently unsupported operation`);
          }}
        >
          Join
        </button>
        <button type="button" onClick={() => { rejectInvite(invite.id); }}>
            Reject
        </button>
      </div>
    </li>
  );
}

const GroupInvites = ({ pendingInvites }: Props): ReactElement | null => {
  if (pendingInvites.isEmpty()) { return null; }
  return (
    <ul className={styles.Banner}>
      {pendingInvites.valueSeq().map((invite) => SingleInvitation(invite))}
    </ul>
  );
};

const Connected = connect(({ pendingInvites }: State) => ({ pendingInvites }))(GroupInvites);
export default Connected;
