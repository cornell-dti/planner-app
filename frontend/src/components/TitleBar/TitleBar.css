.Main {
  margin: 0;
  padding: 32px 32px 16px 32px;
}

.Time {
  font-weight: 600;
  font-size: 32px;
  display: inline-block;
  border-right: 1px black solid;
  padding-right: 10px;
  margin-right: 10px;
  height: 32px;
  line-height: 32px;
  vertical-align: top;
}

.Date {
  line-height: 32px;
  height: 32px;
  display: inline-block;
  vertical-align: top;
}

.Links {
  float: right;
}

.Links button {
  background: none;
  border: none;
  cursor: pointer;
  line-height: 32px;
}
