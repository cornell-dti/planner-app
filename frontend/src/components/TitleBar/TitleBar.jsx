// @flow strict

import React from 'react';
import type { Node } from 'react';
import Clock from 'react-live-clock';
import { date2StringWithYear } from '../../util/datetime-util';
import styles from './TitleBar.css';
import SettingsButton from './Settings/SettingsButton';

export default (): Node => (
  <header className={styles.Main}>
    <span className={styles.Time}><Clock format="h:mm A" ticking /></span>
    <span className={styles.Date}>
      {date2StringWithYear(new Date())}
    </span>
    <span className={styles.Links}><SettingsButton /></span>
  </header>
);
