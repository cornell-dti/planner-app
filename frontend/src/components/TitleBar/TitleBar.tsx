import React, { ReactElement } from 'react';
// @ts-ignore this module does not have a ts definition file.
import Clock from 'react-live-clock';
import { date2FullDateString } from '../../util/datetime-util';
import styles from './TitleBar.css';
import SettingsButton from './Settings/SettingsButton';

/**
 * The title bar.
 *
 * @type {function(): Node}
 */
export default (): ReactElement => (
  <header className={styles.Main}>
    <span className={styles.Time}><Clock format="h:mm A" ticking /></span>
    <span className={styles.Date}>{date2FullDateString(new Date())}</span>
    <span className={styles.Links}><SettingsButton /></span>
  </header>
);
