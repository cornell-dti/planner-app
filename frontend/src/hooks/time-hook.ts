import { useState, useEffect } from 'react';

const INTERVAL = 1000;

let currentTime: number = new Date().getTime();

let testTime: number | null = null;

type Handler = (time: number) => void;

const handlers: Set<Handler> = new Set();

setInterval(() => {
  if (testTime === null) {
    currentTime = new Date().getTime();
    handlers.forEach((handler) => handler(currentTime));
  }
}, INTERVAL);

/**
 * Enable test mode for components that uses or indirectly uses the time hooks.
 * The test mode disables:
 * - Current time real time update.
 * - Reporting updated time to components.
 *
 * @param timeForTesting the value of current time used for deterministic time.
 */
export const enableTestMode = (timeForTesting: number): void => {
  testTime = timeForTesting;
  currentTime = testTime;
};

/**
 * Restart realtime time update when the test ends.
 */
export const disableTestMode = (): void => {
  testTime = null;
};

/**
 * A react state hooks that returns the current time every 1 second.
 * It will cause a re-render when a new value is returned.
 *
 * @returns the current time.
 */
export const useTime = (): number => {
  const [time, setTime] = useState(currentTime);

  useEffect(() => {
    const handler = (newTime: number): void => setTime(newTime);
    handlers.add(handler);
    return (): void => {
      handlers.delete(handler);
    };
  }, []);

  return time;
};

/**
 * A react state hooks that returns the current date every one day.
 * It will cause a re-render when a new value is returned.
 *
 * @returns a date object that falls within today's range.
 */
const useDate = (): Date => {
  const [currentDateString, setCurrentDateString] = useState(new Date(currentTime).toDateString());

  useEffect(() => {
    const handler = (newTime: number): void => {
      const newTimeDateString = new Date(newTime).toDateString();
      if (newTimeDateString !== currentDateString) {
        setCurrentDateString(newTimeDateString);
      }
    };
    handlers.add(handler);
    return (): void => {
      handlers.delete(handler);
    };
  }, [currentDateString]);

  return new Date(currentDateString);
};

/**
 * A react state hooks that returns the current date every one day.
 * It will cause a re-render when a new value is returned.
 *
 * @returns today at 23:59:59 (in user's browser's reported timezone) as a date object.
 */
export const useTodayLastSecondTime = (): Date => {
  const date = new Date(useDate());
  date.setHours(23, 59, 59);
  return date;
};

/**
 * A react state hooks that returns the current date every one day.
 * It will cause a re-render when a new value is returned.
 *
 * @returns today at 00:00:00 (in user's browser's reported timezone) as a date object.
 */
export const useTodayFirstSecondTime = (): Date => {
  const date = new Date(useDate());
  date.setHours(0, 0, 0);
  return date;
};
