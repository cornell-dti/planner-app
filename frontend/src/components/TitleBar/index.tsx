import React, { ReactElement } from 'react';
// @ts-ignore this module does not have a ts definition file.
import Clock from 'react-live-clock';
import Banner from './Banner';
import { date2FullDateString } from '../../util/datetime-util';
import styles from './index.module.css';
import SettingsButton from './Settings/SettingsButton';

/**
 * The title bar.
 *
 * @type {function(): Node}
 */
export default (): ReactElement => (
  <header className={styles.Main}>
    <Banner />
    <span className={styles.Time}><Clock format="h:mm A" ticking /></span>
    <span className={styles.Date}>{date2FullDateString(new Date())}</span>
    <span className={styles.Links}><SettingsButton /></span>
  </header>
);
