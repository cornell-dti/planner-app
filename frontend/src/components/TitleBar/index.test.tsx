import React from 'react';
import renderer from 'react-test-renderer';
import { ProviderForTesting } from 'store';
import { enableTestMode, disableTestMode } from 'hooks/time-hook';
import TitleBar from '.';

it('TitleBar matches snapshot', () => {
  enableTestMode(1800000000000); // around 2027-01-15
  const tree = renderer.create(<ProviderForTesting><TitleBar /></ProviderForTesting>).toJSON();
  expect(tree).toMatchSnapshot();
  disableTestMode();
});
