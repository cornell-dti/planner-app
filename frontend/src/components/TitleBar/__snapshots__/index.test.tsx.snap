// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TitleBar matches snapshot 1`] = `
<header
  className="Main"
>
  <span
    className="Time"
    title="time"
  >
    3:00 AM
  </span>
  <span
    className="Date"
    title="date"
  >
    January 15, 2027
  </span>
  <span
    className="Links"
  >
    <div
      style={
        Object {
          "display": "inline-block",
        }
      }
    >
      <button
        onClick={[Function]}
        type="submit"
      >
        <p
          style={
            Object {
              "transform": "scale(2)translateY(-5px)",
            }
          }
          title="Settings Button"
        >
          <span
            title="Settings"
          >
            <svg
              alt="Settings"
              height="1em"
              onKeyUp={[Function]}
              tabIndex={0}
              title="Settings"
              width="1em"
            >
              settings.svg
            </svg>
          </span>
        </p>
      </button>
    </div>
  </span>
</header>
`;
